File: physicalai-lv1-assignments-main/lv1_module2_student/ros2_ws/src/turtle_cpp/src/distance_subscriber.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float32.hpp"
#include <memory>

class DistanceSubsriber : public rclcpp::Node
{
  public:
    DistanceSubsriber() : Node("distance_subscriber")
    {
        this->declare_parameter("warn_distance", 3.0);

        subscription_ = this->create_subscription<std_msgs::msg::Float32>(
            "turtle_distance", 10, std::bind(&DistanceSubsriber::distance_callback, this, std::placeholders::_1));
    }

  private:
    void distance_callback(const std_msgs::msg::Float32::SharedPtr msg)
    {
        double warn_dist_ = this->get_parameter("warn_distance").as_double();

        float current_dist_ = msg->data;

        if (current_dist_ > warn_dist_)
        {
            RCLCPP_INFO(this->get_logger(), "경고! 임계값 %.1f보다 큽니다", warn_dist_);
        }
    }

    rclcpp::Subscription<std_msgs::msg::Float32>::SharedPtr subscription_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<DistanceSubsriber>();

    rclcpp::spin(node);

    RCLCPP_INFO(node->get_logger(), "DistanceSubscriber 노드 종료");
    rclcpp::shutdown();

    return 0;
}

File: physicalai-lv1-assignments-main/lv1_module2_student/ros2_ws/src/turtle_py/turtle_py/distance_subscriber.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import Float32
from rclpy.signals import SignalHandlerOptions

class DistanceSubscriber(Node):

    def __init__(self):
        super().__init__('distance_subscriber')

        self.declare_parameter('warn_distance', 3.0)

        self.subscription = self.create_subscription(
            Float32,
            'turtle_distance',
            self.distance_callback,
            10
        )

    def distance_callback(self,msg):
        warn_dist = self.get_parameter('warn_distance').value

        current_dist = msg.data

        if current_dist > warn_dist:
            self.get_logger().info(f'경고! 임계값 {warn_dist}보다 큽니다')

def main():
    rclpy.init(signal_handler_options=SignalHandlerOptions.NO)
    node = DistanceSubscriber()

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.get_logger().info('DistanceSubscriber 노드 종료')
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()

if __name__ == '__main__':
    main()